src/libpmemfile-posix/inode.c

```diff
@@ -0,0 +1,841 @@
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, true, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * inode_ref -- increases inode reference counter
+ *
+ * Assumes inode was not allocated in the same transaction.
+ * Return volatile inode.
+ */
+struct pmemfile_vinode *
+inode_ref(PMEMfilepool *pfp,
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, false, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * vinode_tx_unref -- decreases inode reference counter
+ *
+ * Must be called in a transaction.
+ */
+static bool
+vinode_tx_unref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_WORK);
+
+	if (__sync_sub_and_fetch(&vinode->ref, 1) > 0)
+		return false;
+
+	if (vinode->inode->nlink == 0) {
+		inode_array_unregister(pfp, vinode->orphaned.arr,
+				vinode->orphaned.idx);
+
+		inode_free(pfp, vinode->tinode);
+	}
+
+	return true;
+}
+
+/*
+ * vinode_unref -- decreases inode reference counter
+ *
+ * Can't be called in a transaction.
+ */
+void
+vinode_unref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_NONE);
+
+	struct pmemfile_inode_map *c = pfp->inode_map;
+
+	os_rwlock_wrlock(&c->rwlock);
+
+	while (vinode) {
+		struct pmemfile_vinode *to_unregister = NULL;
+		TX_BEGIN_CB(pfp->pop, cb_queue, pfp) {
+			struct pmemfile_vinode *parent = vinode->parent;
+
+			if (vinode_tx_unref(pfp, vinode))
+				to_unregister = vinode;
+
+			if (to_unregister && vinode != pfp->root)
+				vinode = parent;
+			else
+				vinode = NULL;
+		} TX_ONCOMMIT {
+			if (to_unregister)
+				vinode_unregister_locked(pfp, to_unregister);
+		} TX_ONABORT {
+			FATAL("!");
+		} TX_END
+	}
+
+	os_rwlock_unlock(&c->rwlock);
+}
+
+/*
+ * file_get_time -- sets *t to current time
+ */
+void
+file_get_time(struct pmemfile_time *t)
+{
+	pmemfile_timespec_t tm;
+	if (clock_gettime(CLOCK_REALTIME, &tm)) {
+		ERR("!clock_gettime");
+		pmemfile_tx_abort(errno);
+	}
+	t->sec = tm.tv_sec;
+	t->nsec = tm.tv_nsec;
+}
+
+/*
+ * inode_alloc -- allocates inode
+ *
+ * Must be called in a transaction.
+ */
+struct pmemfile_vinode *
+inode_alloc(PMEMfilepool *pfp, uint64_t flags, struct pmemfile_vinode *parent,
+		volatile bool *parent_refed, const char *name, size_t namelen)
+{
+	LOG(LDBG, "flags 0x%lx", flags);
+
+	ASSERTeq(pmemobj_tx_stage(), TX_STAGE_WORK);
+
+	TOID(struct pmemfile_inode) tinode = TX_ZNEW(struct pmemfile_inode);
+	struct pmemfile_inode *inode = D_RW(tinode);
+
+	struct pmemfile_time t;
+	file_get_time(&t);
+
+	inode->version = PMEMFILE_INODE_VERSION(1);
+	inode->flags = flags;
+	inode->ctime = t;
+	inode->mtime = t;
+	inode->atime = t;
+	inode->nlink = 0;
+	os_rwlock_rdlock(&pfp->cred_rwlock);
```

Comment:
IMO cross-module locks (i.e. global locks, locks inside of public structs) should only be used if there are huge performance gains to be had from that solution, otherwise, locks should be as local (in code) as possible.

In this case, we have a fairly small structure that we want to copy/modify. Make a new file with two functions and an internal structure. The structure will contain the lock and the rw instances of the data. The functions will be: cred_read, cred_write. For reads, you either take a pointer or return a copy, and for writes likewise, you get a new copy from the user or a const pointer to new data.

If you find yourself requiring this often, I would probably abstract this into an owning mutex, i.e.: a mutex that internally contains a copy of an arbirary structure, and only exposes safe read/write APIs.